File: backend/app/data/padukuhan.py
"""Keterangan tetap padukuhan: satu baris tunggal di tabel `padukuhan`.

Barisnya BOLEH tidak ada. Selama Admin belum pernah menyimpannya, `ambil()`
mengembalikan `None` dan frontend memakai nilai bawaannya sendiri
(`lib/padukuhan.ts`). Itu disengaja: kalau server ikut menyimpan nilai awal,
ada dua daftar nilai bawaan di dua bahasa yang bisa berbeda diam-diam, dan
tidak ada yang tahu mana yang menang.
"""

from app.core.config import settings
from app.data import db
from app.schemas.padukuhan import Padukuhan

_KOLOM = (
    "nama, namaLengkap, desa, kapanewon, kabupaten, provinsi, luasWilayah,"
    " telepon, email, sejarah, batasUtara, batasTimur, batasSelatan, batasBarat"
)


def ambil() -> Padukuhan | None:
    """Keterangan yang tersimpan, atau `None` kalau Admin belum pernah menyimpan."""
    with db.koneksi(settings.DATABASE_FILE) as conn:
        row = conn.execute(f"SELECT {_KOLOM} FROM padukuhan WHERE id = 1").fetchone()
    return Padukuhan(**dict(row)) if row else None


def ubah(baru: Padukuhan) -> Padukuhan:
    """Simpan seluruh keterangan sekaligus — formnya memang mengirim semuanya,
    jadi tidak ada perpaduan sebagian yang bisa menyisakan kolom setengah lama."""
    kolom = [k.strip() for k in _KOLOM.split(",")]
    with db.koneksi(settings.DATABASE_FILE) as conn:
        conn.execute(
            f"INSERT INTO padukuhan (id, {_KOLOM}) VALUES"
            f" (1, {', '.join(':' + k for k in kolom)})"
            f" ON CONFLICT(id) DO UPDATE SET"
            f" {', '.join(f'{k} = excluded.{k}' for k in kolom)}",
            baru.model_dump(),
        )
        conn.commit()
    return baru


def demo() -> None:
    """Self-check. Jalankan:
    DATABASE_PATH=/tmp/uji-padukuhan.db .venv/bin/python -m app.data.padukuhan
    """
    assert ambil() is None, "DB uji harus mulai tanpa baris padukuhan"

    def contoh(**ganti: str) -> Padukuhan:
        dasar = dict(
            nama="Gading Kulon", namaLengkap="Padukuhan Gading Kulon",
            desa="Donokerto", kapanewon="Kapanewon Turi", kabupaten="Sleman",
            provinsi="Daerah Istimewa Yogyakarta", luasWilayah="162,4 ha",
            telepon="+62 812-2761-391", email="gadingkulon@gmail.com",
            sejarah="Padukuhan di lereng Merapi yang hidup dari pertanian.",
            batasUtara="Gading Lor", batasTimur="Gading Wetan",
            batasSelatan="Ngipak", batasBarat="Banyusoco",
        )
        return Padukuhan(**{**dasar, **ganti})

    assert ubah(contoh()).telepon == "+62 812-2761-391"
    tersimpan = ambil()
    assert tersimpan is not None and tersimpan.desa == "Donokerto"

    # Menyimpan lagi menimpa baris yang sama, bukan menambah baris kedua.
    ubah(contoh(telepon="+62 811-0000-000"))
    with db.koneksi(settings.DATABASE_FILE) as conn:
        (cacah,) = conn.execute("SELECT COUNT(*) FROM padukuhan").fetchone()
    assert cacah == 1, cacah
    assert ambil().telepon == "+62 811-0000-000"

    for salah in ({"nama": "   "}, {"email": "bukan-surel"}, {"sejarah": "pendek"}):
        try:
            contoh(**salah)
        except ValueError:
            pass
        else:
            raise AssertionError(f"seharusnya ditolak: {salah}")

    print("OK: app/data/padukuhan.py")


if __name__ == "__main__":
    demo()

File: frontend/src/lib/padukuhan.ts
export interface Padukuhan {
  nama: string;
  namaLengkap: string;
  desa: string;
  kapanewon: string;
  kabupaten: string;
  provinsi: string;
  luasWilayah: string;

  telepon: string;
  email: string;

  sejarah: string;
  batasUtara: string;
  batasTimur: string;
  batasSelatan: string;
  batasBarat: string;
}

export const PADUKUHAN_BAWAAN: Padukuhan = {
  nama: 'Gading Kulon',
  namaLengkap: 'Padukuhan Gading Kulon',
  desa: 'Donokerto',
  kapanewon: 'Kapanewon Turi',
  kabupaten: 'Sleman',
  provinsi: 'Daerah Istimewa Yogyakarta',
  luasWilayah: '162,4 ha',
  telepon: '+62 812-2761-391',
  email: 'gadingkulon@gmail.com',
  sejarah: [
    'Gading Kulon adalah salah satu padukuhan di Kalurahan Donokerto, Kapanewon Turi, Sleman, Daerah Istimewa Yogyakarta.',
    'Sejak awal berdirinya, warga Gading Kulon hidup produktif berlandaskan kebersamaan, pertanian, dan potensi lokal daerah lereng Gunung Merapi.',
    'Hari ini Gading Kulon berkembang tanpa meninggalkan gotong royong yang jadi wataknya: kerja bakti rutin, ronda malam bergilir, dan kegiatan Karang Taruna yang menggerakkan warga muda. Portal ini dibuat agar data kependudukan padukuhan bisa dibaca dengan cepat, oleh pengurus maupun warga.',
  ].join('\n\n'),
  batasUtara: 'Padukuhan Gading Lor',
  batasTimur: 'Padukuhan Gading Wetan',
  batasSelatan: 'Padukuhan Ngipak',
  batasBarat: 'Kalurahan Banyusoco',
};

export const PETA = {
  koordinat: { lat: -7.656826, lon: 110.363111 },

  radiusPeta: 0.012,
} as const;

export const GOOGLE_MAPS_URL = `https://www.google.com/maps/search/?api=1&query=${PETA.koordinat.lat},${PETA.koordinat.lon}`;

export function paragrafSejarah(sejarah: string): string[] {
  return sejarah
    .split(/\n\s*\n/)
    .map((p) => p.trim())
    .filter(Boolean);
}

export interface BatasWilayah {
  arah: 'Utara' | 'Timur' | 'Selatan' | 'Barat';
  wilayah: string;
}

export function batasWilayah(p: Padukuhan): BatasWilayah[] {
  return [
    { arah: 'Utara', wilayah: p.batasUtara },
    { arah: 'Timur', wilayah: p.batasTimur },
    { arah: 'Selatan', wilayah: p.batasSelatan },
    { arah: 'Barat', wilayah: p.batasBarat },
  ];
}
